### PR TITLE
Story 608

### PR DESCRIPTION
**Pull request by deepakr123**

### CODE REVIEW

File: controllers/v1/assessments.js [hunk 1]
```diff
@@ -835,3 +835,148 @@ async function getUserProfileFunc(req,res){
...
+                    bodyParam.filter.fields[0].dimension = req.body.entityType;
+                    bodyParam.filter.fields[0].value = req.body.entityId;
+                    bodyParam.filter.fields[1].value = req.body.programId;
+                    bodyParam.filter.fields[2].value = req.body.solutionId;
+
+                    let options = config.druid.options;
+                    options.method = "POST";
+                    options.body = bodyParam;
+
+                    let data = await rp(options);
```
Comment (aks30) [inline controllers/v1/assessments.js:960]: @deepakr123 - Use a limit to load only the threshold no of documents in descending order by started date

File: controllers/v1/observations.js [hunk 2]
```diff
@@ -67,35 +67,32 @@ async function instanceObservationData(req, res) {
...
               //if filter is given
-              if (req.body.filter) {
-                if (req.body.filter.questionId && req.body.filter.questionId.length > 0) {
-                  let filter = { "type": "and", "fields": [{ "type": "selector", "dimension": "observationSubmissionId", "value": submissionId }, { "type": "in","dimension":"questionExternalId","values": req.body.filter.questionId }] };
-                  bodyParam.filter = filter;
-                }
-                else {
-                  bodyParam.filter.value = submissionId;
-                }
+              if (req.body.filter && req.body.filter.questionId && req.body.filter.questionId.length > 0 ) {
```
Comment (aks30) [inline controllers/v1/observations.js:87]: @deepakr123 - Optional questions should not be removed from the report and only child dependent questions if not answered we should skip

File: helper/chart_data.js [hunk 3]
```diff
@@ -2263,3 +2263,220 @@ const createSolutionsArray = async function (data) {
...
+
+    return chartObj;
+}
+
+
+//Function to create stacked bar chart response object for entity assessment API
+const entityAssessmentReportChartCreate = async function (domainData) {
+    try {
+
+        let obj = {
```
Comment (aks30) [inline helper/chart_data.js]: @deepakr123 - No of levels are dynamic

File: controllers/v1/observations.js (hunk 2 above)
Comment (deepakr123) [reply controllers/v1/observations.js:87]: resolved

File: helper/chart_data.js (hunk 3 above)
Comment (deepakr123) [reply helper/chart_data.js]: resolved

File: controllers/v1/assessments.js (hunk 1 above)
Comment (deepakr123) [reply controllers/v1/assessments.js:960]: resolved

File: helper/chart_data.js [hunk 4]
```diff
@@ -2263,3 +2263,225 @@ const createSolutionsArray = async function (data) {
...
+   
+      //group the data based on completed date   
+      let groupedSubmissionData = await groupArrayByGivenField(data, "completedDate");
+
+      let completedDate = Object.keys(groupedSubmissionData);
+      
+      //Sort the date in ascending order
+      let sortedDate = completedDate.sort();
+
+      if(completedDate.length > config.druid.no_of_assessment_submissions_threshold) {
```
Comment (aks30) [inline helper/chart_data.js]: @deepakr123 - Define default value if not provided in config

File: helper/chart_data.js [hunk 5]
```diff
@@ -2263,3 +2263,225 @@ const createSolutionsArray = async function (data) {
...
+         completedDate = [sortedDate[0],sortedDate.length-2, sortedDate.length-1];
+      }
+      
+      let domainDataArray = [];
+
+      await Promise.all(completedDate.map(async submission => {
```
Comment (aks30) [inline helper/chart_data.js]: @deepakr123 - check if the result is in a proper sequence

File: helper/chart_data.js (hunk 4 above)
Comment (deepakr123) [reply helper/chart_data.js]: resolved

File: helper/chart_data.js (hunk 5 above)
Comment (deepakr123) [reply helper/chart_data.js]: resolved

File: helper/chart_data.js [hunk 6]
```diff
@@ -2263,3 +2263,255 @@ const createSolutionsArray = async function (data) {
...
+
+        //group the data based on completed date   
+        let groupedSubmissionData = await groupArrayByGivenField(data, "completedDate");
+
+        let completedDateArray = Object.keys(groupedSubmissionData);
+
+        //Sort the date in ascending order
+        completedDateArray = completedDateArray.sort();
+
+        if (completedDateArray.length > config.druid.no_of_assessment_submissions_threshold || 3) {
```
Comment (aks30) [inline helper/chart_data.js]: @deepakr123 - Use constant for defining the default value, check for a valid value in config json i.e. only number is allowed

File: helper/chart_data.js [hunk 7]
```diff
@@ -2263,3 +2263,255 @@ const createSolutionsArray = async function (data) {
...
+            completedDateArray = [sortedDate[0], sortedDate.length - 2, sortedDate.length - 1];
```
Comment (aks30) [inline helper/chart_data.js]: @deepakr123 -Handle the case for submission count being different, i.e. more than 3

File: helper/chart_data.js [hunk 8]
```diff
@@ -2263,3 +2271,296 @@ const createSolutionsArray = async function (data) {
...
+        let sortedData = completedDateKeys.sort();
+
+        let threshold = config.druid.no_of_assessment_submissions_threshold ? config.druid.no_of_assessment_submissions_threshold : default_no_of_assessment_submissions_threshold;
+
+        if(typeof threshold !== "number"){
+            throw new Error("no_of_assessment_submissions_threshold value should be integer");
+        }
+
+        if (sortedData.length > threshold) {
+            console.log("entered");
```
Comment (aks30) [inline helper/chart_data.js]: @deepakr123 - Please remove console

File: helper/chart_data.js [hunk 9]
```diff
@@ -2263,3 +2271,296 @@ const createSolutionsArray = async function (data) {
...
+            sortedData = [sortedData[0], sortedData.length - 2, sortedData.length - 1];
```
Comment (aks30) [inline helper/chart_data.js]: @deepakr123 - Does not work with threshold more than 3

File: helper/chart_data.js (hunk 8 above)
Comment (deepakr123) [reply helper/chart_data.js]: resolved

File: helper/chart_data.js (hunk 6 above)
Comment (deepakr123) [reply helper/chart_data.js]: resolved

File: helper/chart_data.js (hunk 7 above)
Comment (deepakr123) [reply helper/chart_data.js]: resolved

File: helper/chart_data.js (hunk 9 above)
Comment (deepakr123) [reply helper/chart_data.js]: resolved

File: trial.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
+
```
Comment (aks30) [inline trial.js]: @deepakr123 - Please remove this file from git commit

File: helper/chart_data.js [hunk 11]
```diff
@@ -2263,3 +2271,303 @@ const createSolutionsArray = async function (data) {
...
+        if(typeof threshold !== "number"){
+            throw new Error("no_of_assessment_submissions_threshold value should be integer");
+        }
+        
+        let dateArray = [];
+        if (completedDateKeys.length > threshold) {
+
+            dateArray.push(completedDateKeys[0]);
+
+            for (i = completedDateKeys.length - 1; i >= 0; i--) {
```
Comment (aks30) [inline helper/chart_data.js]: @deepakr123 - The current code runs empty for loop cycles. Consider using this - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice

Comment (deepakr123) [reply helper/chart_data.js]: resolved

File: trial.js (hunk 10 above)
Comment (deepakr123) [reply trial.js]: resolved

**Pull request merged**